### PR TITLE
Bot Detection Feature - APIM

### PR DESCRIPTION
**Pull request by nadeepoornima**
**Issue** 
[1]. https://github.com/wso2/product-apim/issues/5358

What have done:

- Has implemented a mediator to detect bot attacks 
- Once detect those details published to Analytics side
- Admin UI to view them and configure emails to get the bot detection alert
- Admin rest API for add, list and delete those configured emails
- New two tables for AM( 1 table) and Analytics(1 table create through the siddhi) side and AM table need to share to Analytics as well
- The required stream has added and modified api-manger.xml to configure that stream
- Modified the log4j to create a separate log file to trace the bit detection

### CODE REVIEW

File: components/apimgt/org.wso2.carbon.apimgt.api/src/main/java/org/wso2/carbon/apimgt/api/model/botDataAPI/BotDetectionData.java [hunk 1]
```diff
@@ -0,0 +1,96 @@
+package org.wso2.carbon.apimgt.api.model.botDataAPI;
+
+public class BotDetectedData {
```
Comment (bhathiya) [inline components/apimgt/org.wso2.carbon.apimgt.api/src/main/java/org/wso2/carbon/apimgt/api/model/botDataAPI/BotDetectionData.java]: Let's rename to "BotDetectionData"

File: components/apimgt/org.wso2.carbon.apimgt.hybrid.gateway/org.wso2.carbon.apimgt.hybrid.gateway.usage.publisher/src/main/java/org/wso2/carbon/apimgt/hybrid/gateway/usage/publisher/APIMgtUsageFileDataPublisher.java [hunk 2]
```diff
@@ -24,14 +24,7 @@
...
 import org.wso2.carbon.apimgt.usage.publisher.DataPublisherUtil;
-import org.wso2.carbon.apimgt.usage.publisher.dto.AlertTypeDTO;
-import org.wso2.carbon.apimgt.usage.publisher.dto.DataBridgeAlertTypesPublisherDTO;
-import org.wso2.carbon.apimgt.usage.publisher.dto.DataBridgeFaultPublisherDTO;
-import org.wso2.carbon.apimgt.usage.publisher.dto.DataBridgeRequestResponseStreamPublisherDTO;
-import org.wso2.carbon.apimgt.usage.publisher.dto.DataBridgeThrottlePublisherDTO;
-import org.wso2.carbon.apimgt.usage.publisher.dto.FaultPublisherDTO;
-import org.wso2.carbon.apimgt.usage.publisher.dto.RequestResponseStreamDTO;
-import org.wso2.carbon.apimgt.usage.publisher.dto.ThrottlePublisherDTO;
+import org.wso2.carbon.apimgt.usage.publisher.dto.*;
```
Comment (bhathiya) [inline components/apimgt/org.wso2.carbon.apimgt.hybrid.gateway/org.wso2.carbon.apimgt.hybrid.gateway.usage.publisher/src/main/java/org/wso2/carbon/apimgt/hybrid/gateway/usage/publisher/APIMgtUsageFileDataPublisher.java]: please revert this change.

File: components/apimgt/org.wso2.carbon.apimgt.hybrid.gateway/org.wso2.carbon.apimgt.hybrid.gateway.usage.publisher/src/main/java/org/wso2/carbon/apimgt/hybrid/gateway/usage/publisher/APIMgtUsageFileDataPublisher.java [hunk 3]
```diff
@@ -131,4 +124,9 @@ public void publishEvent(RequestResponseStreamDTO requestStream) {
             log.error("Error while publishing Request event", e);
         }
     }
+
+    @Override
+    public void publishEvent(BotDataDTO botDataDTO) {
+
+    }
```
Comment (bhathiya) [inline components/apimgt/org.wso2.carbon.apimgt.hybrid.gateway/org.wso2.carbon.apimgt.hybrid.gateway.usage.publisher/src/main/java/org/wso2/carbon/apimgt/hybrid/gateway/usage/publisher/APIMgtUsageFileDataPublisher.java]: why do we need this?

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIAdminImpl.java [hunk 4]
```diff
@@ -158,6 +161,22 @@ public Monetization getMonetizationImplClass() throws APIManagementException {
...
+//    public List<BotDetectedData> getAlert() throws APIManagementException {
+//        return apiMgtDAO.getAlerts();
+//    }
+
+    /**
+     * delete botData api alert my using the messageID
+     */
+//    public void deleteHoneyPotAlert(String messageID) throws APIManagementException, SQLException {
+//        apiMgtDAO.deleteHoneyPotAlert(messageID);
+//    }
```
Comment (bhathiya) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIAdminImpl.java]: should remove these?

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIAdminImpl.java [hunk 5]
```diff
@@ -167,36 +186,50 @@ public MonetizationUsagePublishInfo getMonetizationUsagePublishInfo() throws API
     public void updateMonetizationUsagePublishInfo(MonetizationUsagePublishInfo monetizationUsagePublishInfo)
             throws APIManagementException {
         apiMgtDAO.updateUsagePublishInfo(monetizationUsagePublishInfo);
-    }
+        public void addBotDataEmailConfiguration (String email) throws APIManagementException, SQLException {
+            apiMgtDAO.addBotDataEmailConfiguration(email);
+        }
```
Comment (bhathiya) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIAdminImpl.java]: doesn't this give you a compilation error?

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIAdminImpl.java [hunk 6]
```diff
@@ -167,36 +186,50 @@ public MonetizationUsagePublishInfo getMonetizationUsagePublishInfo() throws API
...
-    public void addMonetizationUsagePublishInfo(MonetizationUsagePublishInfo monetizationUsagePublishInfo)
-            throws APIManagementException {
-        apiMgtDAO.addMonetizationUsagePublishInfo(monetizationUsagePublishInfo);
-    }
+        /**
+         * Add info about monetization usage publish job
+         *
+         * @throws APIManagementException
+         */
+        public static List<BotDetectedData> retrieveSavedBotDataEmailList () throws APIManagementException {
```
Comment (bhathiya) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIAdminImpl.java]: something's wrong here?

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/ApiMgtDAO.java [hunk 7]
```diff
@@ -74,25 +75,28 @@
...
 import org.wso2.carbon.apimgt.impl.factory.SQLConstantManagerFactory;
 import org.wso2.carbon.apimgt.impl.internal.ServiceReferenceHolder;
+<<<<<<< HEAD
 import org.wso2.carbon.apimgt.impl.utils.APIMgtDBUtil;
 import org.wso2.carbon.apimgt.impl.utils.APIUtil;
 import org.wso2.carbon.apimgt.impl.utils.ApplicationUtils;
 import org.wso2.carbon.apimgt.impl.utils.RemoteUserManagerClient;
+=======
+import org.wso2.carbon.apimgt.impl.utils.*;
+>>>>>>> 6a2cb72309... added local part of UI
```
Comment (bhathiya) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/ApiMgtDAO.java]: you have not built the code? :(

File: components/apimgt/org.wso2.carbon.apimgt.api/src/main/java/org/wso2/carbon/apimgt/api/model/botDataAPI/BotDetectedData.java [hunk 8]
```diff
@@ -0,0 +1,100 @@
+package org.wso2.carbon.apimgt.api.model.botDataAPI;
+
+public class BotDetectedData {
```
[the remark this answers is not in the document]
Comment (nadeepoornima) [reply components/apimgt/org.wso2.carbon.apimgt.api/src/main/java/org/wso2/carbon/apimgt/api/model/botDataAPI/BotDetectedData.java]: Chaged the name
